### PR TITLE
feat(crank): convert supporting function-environment-configs

### PR DESCRIPTION
**Pull request by phisco**
### Description of your changes

Now that https://github.com/crossplane/crossplane/pull/5938 was merged, from Crossplane `v1.18.0` native `Composition Environment` was dropped, in favour of [function-environment-configs](https://github.com/crossplane-contrib/function-environment-configs). Hence documentation and tooling is required to help users migrate.

This patch adds a new subcommand `crossplane beta convert composition-environment`, which allows to convert a Composition already in `Pipeline` mode, but still leveraging native `Composition Environment` support, i.e. `spec.environment`, to use `function-environment-configs`.

To also support users still on `Resources` mode Compositions, wanting to migrate to `Pipeline` based compositions on Crossplane 1.18, we need to go through both steps, `Resources` mode -> `Pipeline` mode using `function-patch-and-transform`, but with native Composition Environment, if any -> full Crossplane 1.18 `Pipeline` mode using `function-patch-and-transform` and `function-environment-configs`.

The first transition was covered already by the existing `crossplane beta convert pipeline-composition` subcommand, but its current implementation wasn't preserving unknown fields, going through the `Composition` type it knew, leading to `spec.environment` being dropped altogether, if present, meaning that we couldn't pipe the two commands to get a full transition as needed.

Therefore, I had to rewrite `crossplane beta convert pipeline-composition` to have a more loose `fieldpath`-based approach. While at it I also switched to plain yaml serialization, instead of the schema based, to avoid the issues with creation timestamps. I should have reached similar test coverage as the original implementation, but obviously it's new code and therefore I plan to add more niche corner cases to be sure it behaves correctly.

`crossplane beta convert pipeline-composition` now also goes through the native Composition Environment migration if needed, in order to streamline the experience for users.

So, to be clear, users will only need to run a single command based on their current condition:
- `crossplane beta convert pipeline-composition`: if the Composition is still in `Resource` mode.
- `crossplane beta convert compositione-environment`: if the Composition is already in `Pipeline` mode, but uses other fields in `spec.environment` than `spec.environment.patches`.

I renamed the flag to pass the ref name to a more unique name, so that both the names of function-patch-and-transform and function-environment-configs can be customized. I don't see it as a breaking change being a one-off migration command, worst case we'll invalidate someone's history.

I have: 

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `earthly +reviewable` to ensure this PR is ready for review.
- [x] Added or updated unit tests.
- [ ] ~Added or updated e2e tests.~
- [ ] ~Linked a PR or a [docs tracking issue] to [document this change].~
- [ ] ~Added `backport release-x.y` labels to auto-backport this PR.~

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/main/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/main/contributing#checklist-cheat-sheet

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -26,5 +26,5 @@ gitlab/
 *.xpkg
 
 # go build output (from go build ./cmd/crank etc)
-crank
-crossplane
+/crank
```
Comment (phisco) [inline .gitignore:29]: Had to add this otherwise it was ignoring all changes in cmd/crank... weird

File: cmd/crank/beta/convert/compositionenvironment/cmd.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+// Help returns help message for the migrate composition-environment command.
+func (c *Cmd) Help() string {
+	return `
+This command converts a Crossplane Composition to use function-environment-configs, if needed.
+
+By default it adds a function pipeline step using 
+crossplane-contrib/function-environment-configs, by default it'll reference the function as
+function-environment-configs, but it can be overridden with the -f flag.
```
Comment (lsviben) [inline cmd/crank/beta/convert/compositionenvironment/cmd.go:55]:
```suggestion
It adds a function pipeline step using 
crossplane-contrib/function-environment-configs, by default it'll reference the function as
function-environment-configs, but it can be overridden with the -f flag.
```

File: cmd/crank/beta/convert/compositionenvironment/cmd.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+Examples:
+
+  # Convert an existing Composition to use Pipelines
```
Comment (lsviben) [inline cmd/crank/beta/convert/compositionenvironment/cmd.go:59]:
```suggestion
  # Convert an existing Composition using `spec.environmentConfig` to use function-environment-configs.
```

File: cmd/crank/beta/convert/compositionenvironment/cmd.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	if err := yaml.Unmarshal(data, u); err != nil {
+		return errors.Wrap(err, "Unmarshalling Error")
+	}
+
+	out, err := ConvertToFunctionEnvironmentConfigs(u, c.FunctionEnvironmentConfigRef)
+	if err != nil {
+		return errors.Wrap(err, "Error generating new Composition")
+	}
+	if out == nil {
+		// TODO(phisco): log we didn't do anything maybe
```
Comment (lsviben) [inline cmd/crank/beta/convert/compositionenvironment/cmd.go:96]: 👍 for the log

File: cmd/crank/beta/convert/compositionenvironment/cmd.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+// Run converts a classic Composition to a function pipeline Composition.
+func (c *Cmd) Run() error {
+	data, err := commonIO.Read(c.fs, c.InputFile)
+	if err != nil {
+		return err
+	}
+
+	// Set up schemes for our API types
```
Comment (lsviben) [inline cmd/crank/beta/convert/compositionenvironment/cmd.go:84]: Seems like this is some leftover comment, I dont see any schemas being set up.

File: cmd/crank/beta/convert/compositionenvironment/converter.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	// Copy spec.environment.policy.resolution to function-environment-configs, if any
+	if resolutionPolicy, err := out.GetString("spec.environment.policy.resolution"); err == nil {
+		if err := inputPaved.SetValue("spec.policy.resolution", resolutionPolicy); err != nil {
+			return nil, errors.Wrap(err, "failed to set policy.resolution")
+		}
+		modified = true
+	}
+
```
Comment (lsviben) [inline cmd/crank/beta/convert/compositionenvironment/converter.go:72]: what about `environment.patches`?

File: cmd/crank/beta/convert/compositionenvironment/converter.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+		return nil, nil
+	}
+
+	// Nothing else should be left, we can delete the environment field
+	if err := out.DeleteField("spec.environment"); err != nil {
+		return nil, errors.Wrap(err, "failed to delete environment")
+	}
+
+	// We can drop environmentConfigRefs as well
+	if err := out.DeleteField("spec.environmentConfigRefs"); err != nil && !fieldpath.IsNotFound(err) {
```
Comment (lsviben) [inline cmd/crank/beta/convert/compositionenvironment/converter.go:84]: coulndnt find the field `spec.environmentConfigRefs`, checked older versions. Am I looking at something wrong? 🤔

File: cmd/crank/beta/convert/convert.go [hunk 8]
```diff
@@ -45,5 +47,8 @@ Examples:
   # Convert an existing Composition to use Pipelines
   crossplane beta convert pipeline-composition composition.yaml -o pipeline-composition.yaml
 
+  # Convert an existing Composition to use function-environment-configs instead of native Composition Environment,
+  # requires the composition to be in pipeline mode already.
```
Comment (lsviben) [inline cmd/crank/beta/convert/convert.go:51]: NIT:
```suggestion
  # requires the composition to be in Pipeline mode already.
```

**Review by lsviben — COMMENTED** [review]
Just checked the first part: compositionenvironment

Generally looking good 👍 , added a few comments to clear some things out.

Will do the second part a bit later.

File: cmd/crank/beta/convert/compositionenvironment/converter.go (hunk 6 above)
Comment (phisco) [reply cmd/crank/beta/convert/compositionenvironment/converter.go:72]: if the composition was already moved to `Pipeline` mode, `environmnent.patches` were already moved to `function-patch-and-transform`, so this command doesn't handle that, `crossplane beta convert pipeline-composition` does the full migration from `Resources` to `Pipeline` mode.

File: cmd/crank/beta/convert/compositionenvironment/converter.go (hunk 7 above)
Comment (phisco) [reply cmd/crank/beta/convert/compositionenvironment/converter.go:84]: Good catch, you are right! it's in [composite resources](https://github.com/crossplane/crossplane/blob/79c3d347fbb9646ccfd7cc5e81ee2c936c630413/internal/xcrd/schemas.go#L132).

File: cmd/crank/beta/convert/compositionenvironment/cmd.go (hunk 3 above)
Comment (phisco) [reply cmd/crank/beta/convert/compositionenvironment/cmd.go]: reworded a bit differently, it's not just `spec.environment.environmentConfigs`, there are other fields too, not worth listing them here, I'd just say "using native Composition Environment"

File: cmd/crank/beta/convert/compositionenvironment/cmd.go (hunk 4 above)
Comment (phisco) [reply cmd/crank/beta/convert/compositionenvironment/cmd.go]: added

File: cmd/crank/beta/convert/pipelinecomposition/converter.go [hunk 9]
```diff
@@ -17,186 +17,188 @@ limitations under the License.
...
 	}
-	processedInput.Resources = processedResources
 
-	// Wrap the input in a RawExtension
-	inputType := map[string]any{
-		"apiVersion": "pt.fn.crossplane.io/v1beta1",
-		"kind":       "Resources",
-		"patchSets":  MigratePatchPolicyInPatchSets(processedInput.PatchSets),
-		"resources":  MigratePatchPolicyInResources(processedInput.Resources),
+	// Copy spec.resources, if
```
Comment (lsviben) [inline cmd/crank/beta/convert/pipelinecomposition/converter.go:90]: NIT: part of comment is missing

**Review by lsviben — APPROVED** [review]
Thanks @phisco, LGTM!

Added a few comments, nothing blocking. 

As you mentioned, the changes to pipeline composition are quite extensive, so would be good to run it through a few real world examples. Looks good from what I could tell but there are a lot of pave paths and we may have missed a typo here or there. Although we have unit tests so it makes me more confident.

File: cmd/crank/beta/convert/pipelinecomposition/converter.go [hunk 10]
```diff
@@ -17,186 +17,188 @@ limitations under the License.
...
+	var mode v1.CompositionMode
+	err = out.GetValueInto("spec.mode", &mode)
+	switch {
+	case fieldpath.IsNotFound(err):
+		mode = v1.CompositionModeResources
+	case err != nil:
+		return nil, errors.Wrap(err, "failed to get composition mode")
+	case mode == v1.CompositionModePipeline:
+		// nothing to do
+		return in, nil
```
[the remark this answers is not in the document]
Comment (phisco) [reply cmd/crank/beta/convert/pipelinecomposition/converter.go:65]: Good catch!

File: cmd/crank/beta/convert/pipelinecomposition/cmd.go [hunk 11]
```diff
@@ -71,34 +85,62 @@ func (c *Cmd) AfterApply() error {
...
 
-	oc := &v1.Composition{}
-	_, _, err = decode(data, &v1.CompositionGroupVersionKind, oc)
+	out, err := convertPnTToPipeline(u, c.FunctionPatchAndTransformRef)
 	if err != nil {
-		return errors.Wrap(err, "Decoding Error")
+		return errors.Wrap(err, "Error generating new Composition")
 	}
-
-	_, errs := oc.Validate()
```
[the remark this answers is not in the document]
Comment (phisco) [reply cmd/crank/beta/convert/pipelinecomposition/cmd.go:93]: this was validating the Composition before converting it, I dropped it because it requires unmarshalling into a Composition which would mean dropping the `spec.environment` field right away. Moreover, I'd argue it's not this command's task to validate compositions, it should be just mechanically migrating Compositions independently of their validity. We could validate it after the migration, as it should always produce valid and up-to-date compositions... but how would a user fix an issue on an in-memory Composition? we would need to write it and let the user fix it... but how would they check whether it's valid then? I think we can just keep it simple and avoid validation.

[the remark this answers is not in the document]
Comment (lsviben) [reply cmd/crank/beta/convert/pipelinecomposition/cmd.go:93]: Agree, lets keep it simple and let specified validation tools handle the validation

**Comment by stevendborrelli** [comment]
Initial testing looks good:

- test 1: Pipeline Function already using `function-environment-configs` correctly does not make any changes
- test 2: Pipeline Using built-in `environmentConfig` correctly creates pipeline step for `function-environment-configs` with the same values
- test 3: Legacy P&T in `Resources`  mode: Correctly errors with `Error generating new Composition: Composition is using Resources mode, run pipeline-composition command instead`, converts composition into a Pipeline with `function-environment-configs`

**Review by turkenh — APPROVED** [review]
LGTM, Leaning on detailed reviews/tests from other folks.
